### PR TITLE
Disable renaming for self keyword

### LINKED ISSUE #33902: Disable renaming `self` keyword
**Description:**
$title

**Describe your problem(s)**

**Describe your solution(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by gayalkuruppu**
Fixes #33902 

## Samples
https://user-images.githubusercontent.com/46120162/142967769-93a8e860-eedf-4569-b692-2f363d926eb7.mov

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java [hunk 1]
```diff
@@ -168,7 +177,7 @@ public static WorkspaceEdit rename(RenameContext context) {
         Map<String, ChangeAnnotation> changeAnnotationMap = new HashMap<>();
         WorkspaceEdit workspaceEdit = new WorkspaceEdit();
         Map<String, List<TextEdit>> changes = getChanges(context);
-
+//&& CommonUtil.SELF_KW.equals(context.getParams().getNewName())
```
Comment (TharushiJay) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java:180]:
```suggestion

```

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java [hunk 2]
```diff
@@ -89,7 +90,8 @@ private RenameUtil() {
                 })
                 .orElse(null);
         // Check if token at cursor is an identifier
-        if (!(tokenAtCursor instanceof IdentifierToken) || CommonUtil.isKeyword(tokenAtCursor.text())) {
+        if (!(tokenAtCursor instanceof IdentifierToken) || CommonUtil.isKeyword(tokenAtCursor.text()) 
+                || CommonUtil.SELF_KW.equals(tokenAtCursor.text())) {
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java:94]: As per the language semantics, `self` is not always a keyword. Self is a keyword only within the object constructor expression, and classes. Hence we will need to address that particular scoping check here as well.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java (hunk 2 above)
Comment (IMS94) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java:94]: +1. We can lookup the parents of the `tokenAtCursor` and check if there's a class/object as a parent. Probably move that logic to a separate method.

**Review by mohanvive — DISMISSED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java [hunk 3]
```diff
@@ -128,6 +133,12 @@ private RenameUtil() {
         if (onImportDeclarationNode(context, nodeAtCursor)) {
             return Collections.emptyMap();
         }
+        
+        if (nodeAtCursor.kind() == SyntaxKind.SIMPLE_NAME_REFERENCE) {
+            if (isSelfClassSymbol(context)) {
+                return Collections.emptyMap();
+            }
+        }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/util/rename/RenameUtil.java:142]: Let's merge the two conditions and also, we can merge the particular with the condition at line 133.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

**Review by malinthar — APPROVED** [review]
LGTM

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Pull request merged**